ff/forcefield.py

```diff
@@ -0,0 +1,118 @@
+import ast
+
+class Forcefield():
+
+    def __init__(self, handle):
+        """
+        Initialize the forcefield class.
+
+        Parameters
+        ----------
+        handle: str or dict
+            If str, then the handle is interpret as a path to be opened. If dict,
+            then the handle will be used directly
+        """
+
+        if isinstance(handle, str):
+            handle = open(handle).read()
+            ff_raw = ast.literal_eval(handle)
+        elif isinstance(handle, dict):
+            ff_raw = handle
+
+        global_params = []
+        global_param_groups = []
+
+        def add_param(p, p_group):
+            assert isinstance(p_group, int)
+            length = len(global_params)
+            global_params.append(p)
+            global_param_groups.append(p_group)
+            return length
+
+        # recursively replace parameters with indices and appending them into a global list.
+        def recursive_replace(val, p_group):
+            if isinstance(val, list):
+                arr = []
+                for f_idx, f in enumerate(val):
+                    if isinstance(f, list):
+                        fg = p_group
+                    else:
+                        fg = p_group[f_idx]
+                    v = recursive_replace(f, fg)
+                    arr.append(v)
+                return arr
+            elif isinstance(val, float) or isinstance(val, int):
+                return add_param(val, p_group)
+            else:
+                raise Exception("Unsupported type")
+
+        # (ytz): temporary useful debug code. remove later
+        # print(recursive_replace([0.5, 0.6], (2,3)))
+        # print(recursive_replace([[4.0, 2.0], [5.0, 1.0], [4.0, 233.0], [645.0, 1.0]], (11,12)))
+        # print(global_params)
+        # print(global_param_groups)
+        # assert 0
+
+        group_map = {
+            "Angle": (0,1),
+            "Bond": (2,3),
+            "Improper": (4,5,6),
+            "Proper": (7,8,9),
+            "vdW": (10,11),
+            "GBSA": (12,13),
+            "SimpleCharges": (14,)
+        }
+
```

Comment:
Just to confirm, this indexing will conflict with protein ones right?

Reply:
No - the protein forcefield groups will be the small molecule force groups + 100. I haven't added the protein converter code yet.